File: agent-core/src/api/account.py
"""
api/account.py — Resource Center 账号（登录 / 注册 / 当前身份）。

仪表盘的「我的」界面只跟这里打交道。账号态本身存在浏览器 localStorage 的
`rc_token` 里（bearer token），Agent Core 不持久化任何用户凭据 —— 它只是代理，
避免浏览器直连 Resource Center 时的跨域与自签证书问题。

为什么需要 /session：登录时只拿到一个 token，要显示"我是谁"、以及判断 token
是否已经过期，都得回问 Resource Center。没有这个校验的话，一个过期 token 会让下游每个需要
登录的操作各自回一句 Unauthorized，用户根本不知道该去哪重新登录。
"""

import fastapi
from pydantic import BaseModel
from typing import Optional

router = fastapi.APIRouter(prefix='/account', tags=['account'])


# ── Resource Center 交互 ─────────────────────────────────────────────────────────────────

def _rc_token(request: fastapi.Request) -> Optional[str]:
    """与技能 / 解决方案一致的 X-RC-Token 头。"""
    return request.headers.get('x-rc-token')


async def _rc(method: str, path: str, token: Optional[str] = None,
              json_body=None, timeout: float = 15) -> dict:
    """复用 api/solutions.py 的 Resource Center 请求封装，不再写第二份。"""
    from api.solutions import _rc_request
    return await _rc_request(method, path, token, json_body, timeout)


# Resource Center 的 401 响应体是 `{"ok":false,"error":"Unauthorized"}`。原样透给前端的话，
# 用户看到的就是一个英文单词，既不知道发生了什么也不知道去哪登录。
_UNAUTHORIZED_TEXTS = ('unauthorized', 'not authenticated', 'invalid token')
UNAUTHORIZED_MESSAGE = '未登录或登录已过期，请在「我的」中登录'


def normalize_rc_error(status: int, error: str) -> str:
    """把 Resource Center 的鉴权类报错换成能指向下一步动作的中文。"""
    if status == 401 or (error or '').strip().lower() in _UNAUTHORIZED_TEXTS:
        return UNAUTHORIZED_MESSAGE
    return error or '请求失败'


# ── 端点 ────────────────────────────────────────────────────────────────────

class LoginRequest(BaseModel):
    identifier: str = ''
    password:   str = ''


@router.post('/login')
async def login(req: LoginRequest):
    """登录 Resource Center，返回 bearer token 与身份。"""
    identifier = req.identifier.strip()
    if not identifier or not req.password:
        return {'code': 422, 'error': '请输入账号和密码'}

    result = await _rc('POST', '/api/auth/login', None,
                       {'identifier': identifier, 'password': req.password})
    if not result['ok']:
        # 登录接口的 401 是"账号或密码错误"，不是"登录已过期"，别套用通用文案
        error = result['error']
        if result['status'] == 401:
            error = '账号或密码错误'
        return {'code': result['status'], 'error': error}

    # 登录响应把结果放在顶层：{ok, role, token, userId}
    data = result.get('payload') or {}
    return {'code': 200, 'data': {
        'token':  data.get('token', ''),
        'userId': data.get('userId', ''),
        'role':   data.get('role', ''),
    }}


class RegisterRequest(BaseModel):
    identifier: str = ''
    password:   str = ''
    name:       str = ''


@router.post('/register')
async def register(req: RegisterRequest):
    """注册 Resource Center 账号，成功后直接登录换取 token。

    Resource Center 的注册接口只种 cookie、不发 token（`app/api/auth/register/route.ts`），
    而仪表盘是跨域调用、拿不到那个 cookie，所以注册成功后必须再走一次登录。
    """
    identifier = req.identifier.strip()
    name = req.name.strip()
    if not identifier or not req.password or not name:
        return {'code': 422, 'error': '请输入账号、密码和昵称'}
    if len(req.password) < 8:
        return {'code': 422, 'error': '密码至少 8 位'}

    body = {'identifier': identifier, 'password': req.password, 'name': name}

    result = await _rc('POST', '/api/auth/register', None, body)
    if not result['ok']:
        error = result['error']
        if result['status'] == 409:
            error = '该账号已注册，请直接登录'
        return {'code': result['status'], 'error': error}

    return await login(LoginRequest(identifier=identifier, password=req.password))


@router.get('/session')
async def session(request: fastapi.Request):
    """用浏览器存的 token 问 Resource Center "我是谁"。

    三种结果要分开，否则会误伤：
      200 + user   → 身份有效
      401          → token 真的失效了，前端应该清掉
      其它（含 200 但 ok=false）→ 无法确认。老版本 Resource Center 的 /api/auth/session 只认
        cookie、不认 Bearer，对 token 一律回 `{ok:false,user:null}`；要是把这种
        情况也当成失效，仪表盘在 Resource Center 升级前就永远登不上了。回 503 让前端保留
        token、退化用登录时拿到的身份显示。
    """
    token = _rc_token(request)
    if not token:
        return {'code': 401, 'error': '未登录'}

    result = await _rc('GET', '/api/auth/session', token)
    if not result['ok']:
        if result['status'] == 401:
            return {'code': 401, 'error': UNAUTHORIZED_MESSAGE}
        return {'code': 503,
                'error': f'无法确认登录状态（Resource Center 返回 {result["status"]}）'}

    # session 响应把身份放在顶层的 user 里：{ok, user:{id, role, name, email}}
    user = (result.get('payload') or {}).get('user') or {}
    if not user.get('id'):
        return {'code': 503, 'error': '无法确认登录状态：Resource Center 未返回身份'}
    return {'code': 200, 'data': user}


@router.get('/rc-url')
async def rc_url():
    """当前连的 Resource Center 地址（前端用来给"去网页版注册/管理"的链接）。"""
    from api.skills import _get_rc_url
    return {'code': 200, 'data': {'url': _get_rc_url()}}

File: agent-core/web/js/account.js
/**
 * account.js — 「我的」：Resource Center 账号 + 技能广场 / 解决方案市场入口。
 *
 * 桌面端是顶栏「设置」右边的 modal，移动端是第四个底 tab 的整页面板 ——
 * 两边渲染同一份 HTML（`_render()` 写进两个容器），避免两套逻辑各自漂移。
 *
 * 这里同时是 Resource Center 账号态的唯一出处：token 存在 localStorage.rc_token，
 * skills.js / solutions.js 都从这里取，401 也统一在 `rcFetch` 里处理 ——
 * 以前 Resource Center 的 401 原文（Unauthorized）会被十几处 alert 直接抛给用户，而登录
 * 表单埋在「设置 → 技能 → 我的技能」里，用户根本找不到。
 */

const TOKEN_KEY = 'rc_token';
const ROLE_KEY  = 'rc_role';
const USER_KEY  = 'rc_user_id';

let _overlay, _closeBtn;
let _user = null;         // {id, role, name, email}，未登录为 null
let _mode = 'login';      // login | register
let _notice = '';         // 顶部提示（如"登录已过期"）
let _busy = false;
let _formError = '';
let _rcUrl = '';

// ── Token（唯一出处，skills.js / solutions.js 都 import 这里）──────────────

export function getRcToken() { return localStorage.getItem(TOKEN_KEY) || ''; }

export function setRcToken(token, role, userId) {
  localStorage.setItem(TOKEN_KEY, token);
  if (role) localStorage.setItem(ROLE_KEY, role);
  if (userId) localStorage.setItem(USER_KEY, userId);
}

export function clearRcToken() {
  localStorage.removeItem(TOKEN_KEY);
  localStorage.removeItem(ROLE_KEY);
  localStorage.removeItem(USER_KEY);
  _user = null;
}

/** 登录时记下的身份，用于 Resource Center 暂时问不到 session 时兜底显示。 */
function _storedUser() {
  const id = localStorage.getItem(USER_KEY);
  if (!id) return null;
  return { id, role: localStorage.getItem(ROLE_KEY) || '', name: null, email: null };
}

export function isRcLoggedIn() { return !!getRcToken(); }

/** 带 Resource Center token 的请求头。 */
export function rcHeaders(extra = {}) {
  const h = { 'Content-Type': 'application/json', ...extra };
  const token = getRcToken();
  if (token) h['X-RC-Token'] = token;
  return h;
}

/**
 * 统一的 Resource Center 请求：自动带 token，遇 401 就清 token 并把用户带到「我的」。
 * 返回后端的 JSON（`{code, data|error}`）；401 时也会返回，调用方可以直接
 * 忽略 —— 提示已经由这里给出了。
 */
export async function rcFetch(url, opts = {}) {
  const res = await fetch(url, { ...opts, headers: rcHeaders(opts.headers || {}) });
  let json;
  try { json = await res.json(); } catch { json = { code: res.status, error: '响应解析失败' }; }
  if (json.code === 401 || res.status === 401) {
    const hadToken = isRcLoggedIn();
    clearRcToken();
    showAccount(hadToken ? '登录已过期，请重新登录' : '该操作需要先登录');
  }
  return json;
}

// ── Init ────────────────────────────────────────────────────────────────────

export function initAccount() {
  _overlay  = document.getElementById('account-overlay');
  if (!_overlay) return;
  _closeBtn = document.getElementById('account-close');

  document.getElementById('btn-account').addEventListener('click', () => showAccount());
  _closeBtn.addEventListener('click', hideAccount);
  _overlay.addEventListener('click', (e) => { if (e.target === _overlay) hideAccount(); });

  // 两个容器共用一套事件委托
  ['account-content', 'mobile-account-content'].forEach(id => {
    const el = document.getElementById(id);
    if (el) el.addEventListener('click', _onClick);
    if (el) el.addEventListener('submit', _onSubmit);
  });

  _loadRcUrl();
  refreshAccount();
}

/**
 * 打开「我的」。桌面开 modal；移动端切到第四个 tab（面板本身常驻 DOM）。
 * notice 会显示在顶部，用于"登录已过期"这类被动跳转的说明。
 */
export function showAccount(notice = '') {
  _notice = notice;
  _formError = '';
  const isMobileView = window.matchMedia('(max-width: 768px)').matches;
  if (isMobileView) {
    document.querySelector('.tabbar-btn[data-tab="account"]')?.click();
  } else {
    _overlay?.classList.remove('hidden');
  }
  refreshAccount();
}

export function hideAccount() {
  _overlay?.classList.add('hidden');
  _notice = '';
}

/** 重新拉身份与两个市场的摘要，然后重绘。 */
export async function refreshAccount() {
  _render();                       // 先用现有状态画一次，避免闪空白
  if (getRcToken()) {
    try {
      const res = await fetch('/api/account/session', { headers: rcHeaders() });
      const json = await res.json();
      if (json.code === 200) {
        _user = json.data;
      } else if (json.code === 401) {
        // token 真的失效了：静默回到未登录态。这是"过期 token 到处报
        // Unauthorized"的根治点 —— 在用户动手之前就把状态纠正过来。
        clearRcToken();
        if (!_notice) _notice = '登录已过期，请重新登录';
      } else {
        // 问不到（Resource Center 还没升级到支持 Bearer 的 session、或临时不可用）：
        // 保留 token，用登录时记下的身份兜底，别把用户踢下线。
        _user = _storedUser();
      }
    } catch {
      _user = _storedUser();       // 网络问题同样不清 token
    }
  } else {
    _user = null;
  }
  _syncBadges();
  _render();
  _loadEntrySummaries();
}

async function _loadRcUrl() {
  try {
    const json = await (await fetch('/api/account/rc-url')).json();
    _rcUrl = json.data?.url || '';
  } catch { _rcUrl = ''; }
}

// ── 渲染 ────────────────────────────────────────────────────────────────────

function _render() {
  const html = `
    ${_notice ? `<div class="account-notice">${_esc(_notice)}</div>` : ''}
    ${_user ? _renderIdentity() : _renderAuthForm()}
    ${_renderTokenSection()}
    <div class="account-section-label">内容</div>
    <div class="settings-list account-entry-list">
      <button class="settings-item" data-entry="skills">
        <span class="settings-item-text">
          <span class="settings-item-label">技能广场</span>
          <span class="settings-item-sub" id="account-skills-sub">安装、激活、发布技能</span>
        </span>
        <span class="settings-item-arrow">›</span>
      </button>
      <button class="settings-item" data-entry="solutions">
        <span class="settings-item-text">
          <span class="settings-item-label">解决方案市场</span>
          <span class="settings-item-sub" id="account-solutions-sub">整套画布 / 技能 / Prompt / 任务</span>
        </span>
        <span class="settings-item-arrow">›</span>
      </button>
    </div>
    <p class="account-hint">浏览、安装、加载技能与解决方案不需要登录；发布、管理自己的内容才需要。</p>`;

  ['account-content', 'mobile-account-content'].forEach(id => {
    const el = document.getElementById(id);
    if (el) el.innerHTML = html;
  });
}

function _renderIdentity() {
  const label = _user.email || _user.name || _user.id;
  // id 是 cuid，窄屏上整串会把"复制"挤到第二行；只显示头尾，完整值给 title 与复制
  const shortId = _user.id.length > 12
    ? `${_user.id.slice(0, 6)}…${_user.id.slice(-4)}`
    : _user.id;
  return `
    <div class="account-card account-card-signed">
      <div class="account-avatar">${_esc((label || '?').slice(0, 1).toUpperCase())}</div>
      <div class="account-identity">
        <div class="account-identity-name">${_esc(label)}</div>
        <div class="account-identity-meta">
          <span class="account-role">${_esc(_user.role || 'viewer')}</span>
          <code class="account-id" title="${_esc(_user.id)}">${_esc(shortId)}</code>
          <button class="account-copy-btn" data-copy="${_esc(_user.id)}" title="复制完整 ID">复制</button>
        </div>
      </div>
      <button class="skill-btn skill-btn-sm" data-action="logout">退出登录</button>
    </div>`;
}

/** 登录后才有 token；单独一个 section，别再挤进 identity 卡片的元信息行里。 */
function _renderTokenSection() {
  const token = getRcToken();
  if (!_user || !token) return '';
  const shortToken = token.length > 20
    ? `${token.slice(0, 10)}…${token.slice(-6)}`
    : token;
  return `
    <div class="account-section-label">本机 Token</div>
    <div class="account-card account-token-card">
      <code class="account-token-value" title="${_esc(token)}">${_esc(shortToken)}</code>
      <div class="account-token-actions">
        <button class="account-copy-btn" data-copy="${_esc(token)}" title="复制完整 Token">复制</button>
        <button class="account-link account-token-reset" data-action="reset-token" title="清除本机 Token 并重新登录">重置</button>
      </div>
    </div>`;
}

function _renderAuthForm() {
  const isRegister = _mode === 'register';
  return `
    <div class="account-card">
      <div class="account-card-head">
        <div class="account-avatar account-avatar-empty">?</div>
        <div>
          <div class="account-identity-name">未登录</div>
          <div class="account-identity-sub">登录后可发布技能与解决方案</div>
        </div>
      </div>
      <form class="account-form" data-form="${isRegister ? 'register' : 'login'}">
        <input type="text" name="identifier" placeholder="邮箱账号" autocomplete="username" required>
        <input type="password" name="password" placeholder="${isRegister ? '密码（至少 8 位）' : '密码'}"
               autocomplete="${isRegister ? 'new-password' : 'current-password'}" required
               ${isRegister ? 'minlength="8"' : ''}>
        ${isRegister ? `<input type="text" name="name" placeholder="昵称" autocomplete="nickname" required>` : ''}
        ${_formError ? `<p class="account-form-error">${_esc(_formError)}</p>` : ''}
        <button type="submit" class="account-submit-btn" ${_busy ? 'disabled' : ''}>
          ${_busy ? '处理中…' : (isRegister ? '注册并登录' : '登录')}
        </button>
      </form>
      <div class="account-switch">
        ${isRegister
          ? `<span>已有账号？<button class="account-link" data-action="to-login">去登录</button></span>`
          : `<span>还没有账号？<button class="account-link" data-action="to-register">注册</button></span>`}
      </div>
      ${_rcUrl ? `
        <div class="account-web-link">
          <a class="account-link" href="${_esc(_rcUrl)}" target="_blank" rel="noreferrer">Resource Center 网页版 ↗</a>
        </div>` : ''}
    </div>`;
}

/** 两个入口的副文本：已装技能数 / 当前载入的方案。 */
async function _loadEntrySummaries() {
  try {
    const skills = (await (await fetch('/api/skills')).json()).data || [];
    const active = skills.filter(s => s.active).length;
    _setText('account-skills-sub',
      skills.length ? `已安装 ${skills.length} · 激活 ${active}` : '还没有安装技能');
  } catch { /* 保持默认文案 */ }

  try {
    const current = (await (await fetch('/api/solutions/current')).json()).data;
    _setText('account-solutions-sub', current && (current.name || current.slug)
      ? `当前：${current.name || current.slug}${current.version ? ' v' + current.version : ''}`
      : '尚未载入解决方案');
  } catch { /* 保持默认文案 */ }
}

/** 顶栏按钮与移动端 tab 上的账号角标。 */
function _syncBadges() {
  const label = _user ? (_user.email || _user.name || _user.id).split('@')[0] : '';
  const badge = document.getElementById('account-badge');
  if (badge) {
    badge.textContent = label;
    badge.classList.toggle('hidden', !label);
  }
  // 移动端 tab 上只用一个圆点表示"已登录"，标签放不下文字
  const tabBadge = document.getElementById('account-tab-badge');
  if (tabBadge) {
    tabBadge.textContent = '';
    tabBadge.classList.toggle('hidden', !_user);
  }
}

// ── 交互 ────────────────────────────────────────────────────────────────────

function _onClick(e) {
  const entry = e.target.closest('[data-entry]');
  if (entry) {
    // 两个市场都不需要登录就能浏览，直接点隐藏的原按钮，沿用它们自己的初始化
    hideAccount();
    document.getElementById(entry.dataset.entry === 'skills' ? 'btn-skills' : 'btn-solutions')?.click();
    return;
  }

  const copyBtn = e.target.closest('[data-copy]');
  if (copyBtn) {
    navigator.clipboard?.writeText(copyBtn.dataset.copy);
    copyBtn.textContent = '已复制';
    setTimeout(() => { copyBtn.textContent = '复制'; }, 1500);
    return;
  }

  const action = e.target.closest('[data-action]')?.dataset.action;
  if (action === 'logout') {
    clearRcToken();
    _notice = '';
    _mode = 'login';
    refreshAccount();
  } else if (action === 'reset-token') {
    // Token 是 Resource Center 签发的无状态 JWT，服务端不持有可撤销的会话记录，
    // 所以"重置"只能清掉本机存的这一份——下次操作会引导重新登录换取新 token。
    if (!confirm('重置后需要重新登录才能获得新的 Token，确定继续？')) return;
    clearRcToken();
    _notice = '';
    _mode = 'login';
    refreshAccount();
  } else if (action === 'to-register') {
    _mode = 'register'; _formError = ''; _render();
  } else if (action === 'to-login') {
    _mode = 'login'; _formError = ''; _render();
  }
}

async function _onSubmit(e) {
  const form = e.target.closest('[data-form]');
  if (!form) return;
  e.preventDefault();
  if (_busy) return;

  const kind = form.dataset.form;   // login | register
  const body = {
    identifier: form.identifier.value.trim(),
    password:   form.password.value,
  };
  if (kind === 'register') {
    body.name = form.name.value.trim();
    if (!body.name) {
      _formError = '请输入昵称';
      _render();
      return;
    }
  }

  _busy = true; _formError = ''; _render();
  try {
    const res = await fetch(`/api/account/${kind}`, {
      method: 'POST', headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
    const json = await res.json();
    if (json.code !== 200 || !json.data?.token) {
      _formError = json.error || (kind === 'register' ? '注册失败' : '登录失败');
      _busy = false; _render();
      return;
    }
    setRcToken(json.data.token, json.data.role, json.data.userId);
    _notice = '';
    _mode = 'login';
    _busy = false;
    await refreshAccount();
  } catch (err) {
    _formError = `网络错误：${err.message}`;
    _busy = false;
    _render();
  }
}

// ── Util ────────────────────────────────────────────────────────────────────

function _setText(id, text) {
  const el = document.getElementById(id);
  if (el) el.textContent = text;
}

function _esc(str) {
  if (!str) return '';
  return String(str).replace(/&/g, '&amp;').replace(/</g, '&lt;')
                    .replace(/>/g, '&gt;').replace(/"/g, '&quot;');
}
